mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/SmartTestingMavenConfigurer.java

```diff
@@ -133,16 +139,56 @@ private void calculateChanges(File projectDirectory, Configuration configuration
     }
 
     private void configureExtension(MavenSession session, Configuration configuration) {
-        final MavenProjectConfigurator mavenProjectConfigurator = new MavenProjectConfigurator(configuration);
-        session.getAllProjects().forEach(mavenProject -> {
-            boolean wasConfigured = mavenProjectConfigurator.configureTestRunner(mavenProject.getModel());
-            if (wasConfigured) {
-                configuration.dump(mavenProject.getBasedir());
-                if (isFailedStrategyUsed()) {
-                    SurefireReportStorage.copySurefireReports(mavenProject.getModel());
-                }
+        final Consumer<MavenProject> configureSmartTestingExtensionAction;
+        final File executionRootDirectory = new File(session.getExecutionRootDirectory());
+        if (hasModuleSpecificConfigurations(executionRootDirectory, this::isProjectRootDirectory)) {
+            configureSmartTestingExtensionAction = applyModuleSpecificConfiguration(configuration);
+        } else {
+            final MavenProjectConfigurator mavenProjectConfigurator = new MavenProjectConfigurator(configuration);
+            configureSmartTestingExtensionAction = mavenProject -> configureMavenProject(mavenProjectConfigurator, mavenProject, configuration);
+        }
+        session.getAllProjects().forEach(configureSmartTestingExtensionAction);
+    }
+
+    private boolean hasModuleSpecificConfigurations(File projectDir, Function<File, Boolean> stopCondition) {
+        final ConfigLookup lookUp = new ConfigLookup(projectDir, stopCondition);
+        final boolean moreThanOneConfigFile = lookUp.hasMoreThanOneConfigFile(SMART_TESTING_YML, SMART_TESTING_YAML);
+        return moreThanOneConfigFile && System.getProperty(SMART_TESTING_CONFIG) == null;
+    }
+
+    private Consumer<MavenProject> applyModuleSpecificConfiguration(Configuration configuration) {
+        return mavenProject -> {
+            final ConfigLookup configLookup = new ConfigLookup(mavenProject.getBasedir(), this::isProjectRootDirectory);
+            final File dirWithConfig = configLookup.getFirstDirWithConfigOrWithStopCondition();
+            final Configuration mavenProjectConfiguration = configLookup.isConfigFromProjectRootDir() ? configuration :
```

Comment:
On the previous line, you call `configLookup.getFirstDirWithConfigOrWithStopCondition()` and here using the method `configLookup.isConfigFromProjectRootDir()` you [call](https://github.com/arquillian/smart-testing/pull/262/files#diff-0d186905ac15e0569433af7b3e556ccbR51) the same method again: `getFirstDirWithConfigOrWithStopCondition()`
To be honest, this is very cryptic to my taste.

Reply:
I'm asking myself, is it really worth having this complicated check there? When looking at the code, it doesn't seem to be an optimization at all - this solution is maybe more time and resource consuming.

Reply:
ok let me see how can we make it more readable & less complicated